sympy/core/operations.py

```diff
@@ -490,3 +492,66 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(Dispatcher):
+    """
+    Dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    This class lets the user to add or multiply different types of
+    arguments. It is recommended to define the combination by registering
+    with this class rather than complicating `Add.flatten` and `Mul.flatten`.
+
+    Examples
+    ========
+
+    >>> from sympy import Mul, Expr, Symbol
+    >>> from sympy.vector import CoordSys3D, Vector, VectorMul
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> a, i = Symbol('a'), CoordSys3D('C').base_vectors()[0]
+
+    >>> mul = AssocOpDispatcher('mul', Mul)
+
+    >>> @mul.register(Expr, Vector)
+    ... @mul.register(Vector, Expr)
+    ... @mul.register(Vector, Vector)
+    ... def _(a,b, **kwargs):
+    ...     return VectorMul(a, b, **kwargs)
+
+    >>> mul()
+    1
+    >>> mul(a)
+    a
+    >>> mul(i)
+    C.i
+
+    >>> mul(a, i)
+    a*C.i
+    >>> mul(i, a)
+    a*C.i
+    >>> mul(a, i, a)
+    a**2*C.i
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.fallback = fallback
+        super().__init__(name, doc)
+
+    def __call__(self, *args, **kwargs):
+
+        if len(args) < 2:
+            try:
+                return super().__call__(*args, **kwargs)
+            except NotImplementedError:
+                return self.fallback(*args, **kwargs)
+
+        elif len(args) == 2:
+            try:
+                return super().__call__(*args, **kwargs)
+            except NotImplementedError:
+                return self.fallback(*args, **kwargs)
+
+        else:
+            return functools.reduce(lambda result, a: self(result, a, **kwargs), args)
```

Comment:
Does this cause `flatten` to be called once for each term in the `add` or `mul` call?

The flatten function is carefully designed to be `O(N)` in the the number `N` of terms. However it is necessary to pass all args at once in a single call to `Add/Mul` or it can easily become `O(N**2)`.

Reply:
Abandoning binary dispatch.